File: packages/verify-receipt/src/verify-proof.ts
#!/usr/bin/env node
/**
 * aqta-verify-proof
 *
 * Checks a transparency proof offline. Separate from aqta-verify-receipt
 * because the questions are different: a receipt asks "did the issuer assert
 * this", a proof asks "is this entry in the issuer's log, and has that log
 * only ever grown". Conflating them into one command invites a reader to think
 * a valid signature answered the second question. It does not.
 *
 *   aqta-verify-proof inclusion.json
 *   aqta-verify-proof consistency.json
 *   aqta-verify-proof sth.json --key <published key>
 *   curl -s https://api.aqta.ai/v1/public/transparency/proof/<id> | aqta-verify-proof -
 *
 * Exit 0 valid, 1 invalid, 2 usage or IO.
 */
import { readFileSync } from 'fs';
import {
  verifyConsistencyProof,
  verifyInclusionProof,
  verifySignedTreeHead,
} from './transparency.js';

const USAGE = `aqta-verify-proof <file|-> [--key <base64url>] [--json] [-q]

  Detects the document type from its fields:
    audit_path        an RFC 6962 inclusion proof
    consistency_path  an RFC 6962 consistency proof
    signature         a signed tree head, which needs --key

  A proof establishes that what you were shown is in the log. It does not
  establish that what you were not shown is irrelevant.
`;

function main(): void {
  const args = process.argv.slice(2);
  let file = '';
  let key: string | undefined;
  let asJson = false;
  let quiet = false;

  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === '--key') key = args[++i];
    else if (a === '--json') asJson = true;
    else if (a === '-q' || a === '--quiet') quiet = true;
    else if (a === '-h' || a === '--help') {
      process.stdout.write(USAGE);
      process.exit(0);
    } else if (!file) file = a;
    else {
      process.stderr.write(`aqta-verify-proof: unexpected argument ${a}\n`);
      process.exit(2);
    }
  }
  if (!file) {
    process.stderr.write(USAGE);
    process.exit(2);
  }

  let raw: string;
  try {
    raw = file === '-' ? readFileSync(0, 'utf8') : readFileSync(file, 'utf8');
  } catch {
    process.stderr.write(`aqta-verify-proof: cannot read ${file === '-' ? 'stdin' : file}\n`);
    process.exit(2);
  }

  let doc: Record<string, unknown>;
  try {
    doc = JSON.parse(raw);
  } catch {
    process.stderr.write('aqta-verify-proof: input is not valid JSON\n');
    process.exit(2);
  }
  if (typeof doc !== 'object' || doc === null || Array.isArray(doc)) {
    process.stderr.write('aqta-verify-proof: input must be a JSON object\n');
    process.exit(2);
  }

  // Some endpoints wrap the proof in an envelope; accept either shape.
  const inner = (doc.proof ?? doc.inclusion_proof ?? doc.consistency_proof ?? doc) as Record<string, unknown>;

  let kind: string;
  let result: { valid: boolean; reason?: string };

  if ('audit_path' in inner) {
    kind = 'inclusion proof';
    result = verifyInclusionProof(inner);
  } else if ('consistency_path' in inner) {
    kind = 'consistency proof';
    result = verifyConsistencyProof(inner);
  } else if ('signature' in inner && 'root_hash' in inner) {
    kind = 'signed tree head';
    if (!key) {
      process.stderr.write('aqta-verify-proof: a signed tree head needs --key <published key>\n');
      process.exit(2);
    }
    result = verifySignedTreeHead(inner, key);
  } else {
    process.stderr.write(
      'aqta-verify-proof: unrecognised document: expected audit_path, consistency_path, or a signed tree head\n'
    );
    process.exit(2);
  }

  if (!quiet) {
    if (asJson) {
      process.stdout.write(
        JSON.stringify({
          valid: result.valid,
          kind,
          reason: result.valid ? null : result.reason ?? 'verification failed',
        }) + '\n'
      );
    } else {
      const mark = result.valid ? '✓ valid' : '✕ invalid';
      const detail = result.valid ? kind : `${kind}: ${result.reason ?? 'verification failed'}`;
      process.stdout.write(`${mark}  ${detail}\n`);
    }
  }
  process.exit(result.valid ? 0 : 1);
}

main();

File: packages/verify-receipt-py/src/aqta_verify_receipt/verify_proof.py
"""aqta-verify-proof: check a transparency proof offline.

Separate from `aqta-verify-receipt` because the questions are different. A
receipt asks "did the issuer assert this". A proof asks "is this entry in the
issuer's log, and has that log only ever grown". Conflating them into one
command invites a reader to think a valid signature answered the second
question. It does not.

    aqta-verify-proof inclusion.json
    aqta-verify-proof consistency.json
    aqta-verify-proof sth.json --key <published key>
    curl -s https://api.aqta.ai/v1/public/transparency/proof/<id> | aqta-verify-proof -

Verdicts, exit codes and the document sniffing match the TypeScript command of
the same name byte for byte, because a bounty class nobody outside can exercise
in both languages is not a bounty class.

Exit 0 valid, 1 invalid, 2 usage or IO.
"""
from __future__ import annotations

import json
import sys
from typing import Any, Mapping, Optional

from .transparency import (
    verify_consistency_proof,
    verify_inclusion_proof,
    verify_signed_tree_head,
)

USAGE = """aqta-verify-proof <file|-> [--key <base64url>] [--json] [-q]

  Detects the document type from its fields:
    audit_path        an RFC 6962 inclusion proof
    consistency_path  an RFC 6962 consistency proof
    signature         a signed tree head, which needs --key

  A proof establishes that what you were shown is in the log. It does not
  establish that what you were not shown is irrelevant.
"""


def main(argv: Optional[list] = None) -> None:
    args = list(sys.argv[1:] if argv is None else argv)
    file = ""
    key: Optional[str] = None
    as_json = False
    quiet = False

    i = 0
    while i < len(args):
        a = args[i]
        if a == "--key":
            i += 1
            key = args[i] if i < len(args) else None
        elif a == "--json":
            as_json = True
        elif a in ("-q", "--quiet"):
            quiet = True
        elif a in ("-h", "--help"):
            sys.stdout.write(USAGE)
            raise SystemExit(0)
        elif not file:
            file = a
        else:
            sys.stderr.write(f"aqta-verify-proof: unexpected argument {a}\n")
            raise SystemExit(2)
        i += 1

    if not file:
        sys.stderr.write(USAGE)
        raise SystemExit(2)

    try:
        raw = sys.stdin.read() if file == "-" else open(file, "r", encoding="utf-8").read()
    except OSError:
        where = "stdin" if file == "-" else file
        sys.stderr.write(f"aqta-verify-proof: cannot read {where}\n")
        raise SystemExit(2)

    try:
        doc: Any = json.loads(raw)
    except ValueError:
        sys.stderr.write("aqta-verify-proof: input is not valid JSON\n")
        raise SystemExit(2)
    if not isinstance(doc, Mapping):
        sys.stderr.write("aqta-verify-proof: input must be a JSON object\n")
        raise SystemExit(2)

    # Some endpoints wrap the proof in an envelope; accept either shape.
    inner = doc.get("proof") or doc.get("inclusion_proof") or doc.get("consistency_proof") or doc
    if not isinstance(inner, Mapping):
        sys.stderr.write("aqta-verify-proof: input must be a JSON object\n")
        raise SystemExit(2)

    if "audit_path" in inner:
        kind = "inclusion proof"
        result = verify_inclusion_proof(inner)
    elif "consistency_path" in inner:
        kind = "consistency proof"
        result = verify_consistency_proof(inner)
    elif "signature" in inner and "root_hash" in inner:
        kind = "signed tree head"
        if not key:
            sys.stderr.write(
                "aqta-verify-proof: a signed tree head needs --key <published key>\n"
            )
            raise SystemExit(2)
        result = verify_signed_tree_head(inner, key)
    else:
        sys.stderr.write(
            "aqta-verify-proof: unrecognised document: expected audit_path, "
            "consistency_path, or a signed tree head\n"
        )
        raise SystemExit(2)

    if not quiet:
        if as_json:
            sys.stdout.write(
                json.dumps(
                    {
                        "valid": result.valid,
                        "kind": kind,
                        "reason": None if result.valid else (result.reason or "verification failed"),
                    },
                    # Compact separators so this is byte-identical to the
                    # TypeScript command's JSON.stringify output. Two commands
                    # that disagree on whitespace are two commands to diff.
                    separators=(",", ":"),
                )
                + "\n"
            )
        else:
            mark = "✓ valid" if result.valid else "✕ invalid"
            detail = kind if result.valid else f"{kind}: {result.reason or 'verification failed'}"
            sys.stdout.write(f"{mark}  {detail}\n")

    raise SystemExit(0 if result.valid else 1)


if __name__ == "__main__":
    main()
